presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java

```diff
@@ -568,6 +577,31 @@ public Boolean visitParameter(Parameter node, Void context)
             return process(parameters.get(node.getPosition()), context);
         }
 
+        public Boolean visitGroupingOperation(GroupingOperation node, Void context)
+        {
+            // ensure that no output fields are referenced from ORDER BY clause
+            if (orderByScope.isPresent()) {
+                node.getGroupingColumns().forEach(groupingColumn -> verifyNoOrderByReferencesToOutputColumns(
+                        groupingColumn,
+                        REFERENCE_TO_OUTPUT_ATTRIBUTE_WITHIN_ORDER_BY_GROUPING,
+                        "Invalid reference to output projection attribute from ORDER BY grouping"
+                ));
+            }
+
+            Optional<Expression> argumentNotInGroupBy = node.getGroupingColumns().stream()
+                    .filter(argument -> !expressions.contains(argument))
+                    .findAny();
```

Comment:
(continuing comment in the grammar)

You can automatically support `grouping(t.a) ... group by a` (and vice versa), if you code the above stream like this:

```
Optional<Expression> argumentNotInGroupBy = node.getGroupingColumns().stream()
        .filter(argument ->
                !columnReferences.containsKey(argument)
                        || !isGroupingKey(argument))
        .findAny();
```

(of course, this requires adding a test)